### PR TITLE
mpir/pmi: Fix singleton init with PMI-2 and PMIx

### PR DESCRIPTION
**Pull request by raffenet**
## Pull Request Description

Allow simple programs to run without a PMI-2 or PMIx server (e.g. `./examples/cpi`). Full-fledged singleton init that starts a server is not yet supported with PMI-2 or PMIx.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
This only fixes for PMIx, right? Is it easy to patch PMI-2 as well?

**Comment by raffenet** [comment]
> This only fixes for PMIx, right? Is it easy to patch PMI-2 as well?

I've got about 90% of the fix for PMI-2 done. Should be able to add it here shortly.

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
Rebased to clear the warnings failure.
test:mpich/pmi

**Comment by raffenet** [comment]
Locally testing `./examples/cpi` with both pmi2 and pmix configurations works after these changes.

**Comment by raffenet** [comment]
> Locally testing `./examples/cpi` with both pmi2 and pmix configurations works after these changes.

Going to adjust the build_nodemap changes to try and simplify this some more, especially in the pmi2 case.

**Comment by raffenet** [comment]
> > Locally testing `./examples/cpi` with both pmi2 and pmix configurations works after these changes.
> 
> Going to adjust the build_nodemap changes to try and simplify this some more, especially in the pmi2 case.

I somehow missed that the single process case for building the nodemap is already handled, no matter the pmi configuration. The MPIR_pmi layer calls `build_nodemap_nolocal`. As long as size and rank are set (fixed in [35f5a2a]) then no other changes are needed.

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

File: src/util/mpir_pmi.c [hunk 1]
```diff
@@ -415,8 +415,10 @@ int MPIR_pmi_barrier(void)
     MPIR_ERR_CHKANDJUMP1(pmi_errno != PMI2_SUCCESS, mpi_errno, MPI_ERR_OTHER,
                          "**pmi_kvsfence", "**pmi_kvsfence %d", pmi_errno);
     /* Get a non-existent key, it only returns after every process called fence */
-    int out_len;
-    PMI2_KVS_Get(pmi_jobid, PMI2_ID_NULL, "-NONEXIST-KEY", NULL, 0, &out_len);
+    if (MPIR_Process.size > 1) {
```
Comment (hzhou) [inline src/util/mpir_pmi.c:418]: Can't we skip the whole function when size is 1?

Comment (raffenet) [reply src/util/mpir_pmi.c:418]: Yes, I think that should be OK. I'll update.

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

File: src/pmi/src/pmi_v2.c [hunk 2]
```diff
@@ -365,9 +371,11 @@ PMI_API_PUBLIC int PMI2_KVS_Fence(void)
...
     PMIU_cmd_init(&pmicmd, USE_WIRE_VER, "kvs-fence");
 
-    pmi_errno = PMIU_cmd_get_response(PMI_fd, &pmicmd, "kvs-fence-response");
-    PMIU_ERR_POP(pmi_errno);
-    PMI2_CHK_RC_ERRMSG(&pmicmd);
+    if (PMI_initialized > SINGLETON_INIT_BUT_NO_PM) {
+        pmi_errno = PMIU_cmd_get_response(PMI_fd, &pmicmd, "kvs-fence-response");
+        PMIU_ERR_POP(pmi_errno);
+        PMI2_CHK_RC_ERRMSG(&pmicmd);
+    }
```
Comment (hzhou) [inline src/pmi/src/pmi_v2.c:378]: I believe this is no longer necessary now that we skips barrier in single process?

**Review by hzhou — APPROVED** [review]
Looks good. Just a minor comment --

**Comment by raffenet** [comment]
It looks like skipping the barrier in PMI-1 causes an issue in ch3 where previously put "connInfo" values are not visible to PMI_KVS_Get. We can probably fix the usage to only get values not put by the local process. Will take a look at fixing...

File: src/pmi/src/pmi_v2.c (hunk 2 above)
Comment (raffenet) [reply src/pmi/src/pmi_v2.c:378]: Yes, we can remove this change.

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
I added a singleton `cpi` execution to the test script for `mpich-review-pmi` Jenkins jobs.
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
PMIx singletons in Jenkins are broken. The Jenkins env is using a much older PMIx release than the one I tested with. I'm going to try and get it upgraded to something more recent, and preferably running with prrte instead of Open MPI's mpiexec for testing.

**Comment by raffenet** [comment]
test:mpich/pmi

**Comment by raffenet** [comment]
I think all remaining failures are resolved. One more test run...
test:mpich/pmi

**Review by hzhou — APPROVED** [review]

**Pull request merged**